### PR TITLE
Ledger Integration

### PR DESCRIPTION
**Pull request by SebastienGllmt**
Add ledger support 

(very much a WIP. Just creating the PR so we can all track it)

### CODE REVIEW

File: app/api/ada/index.js [hunk 1]
```diff
@@ -557,12 +624,12 @@ export default class AdaApi {
...
-    request: CreateTrezorWalletRequest
-  ): Promise<CreateTrezorWalletResponse> {
+  async createHardwareWallet(
+    request: CreateHardwareWalletRequest
+  ): Promise<CreateHardwareWalletResponse> {
     try {
-      Logger.debug('AdaApi::connectTrezor called');
-      const { walletName, publicMasterKey, deviceFeatures } = request;
+      Logger.debug('AdaApi::createHardwareWallet called');
+      const { walletName, walletHardwareInfo } = request;
```
Comment (nicarq) [inline app/api/ada/index.js]: it’s not so obvious that contains the `publicMasterKey`. it should be extracted as well and then put back together in `cwHardwareInfo`

File: app/api/common.js [hunk 2]
```diff
@@ -4,12 +4,12 @@
 
 import BigNumber from 'bignumber.js';
 import { defineMessages } from 'react-intl';
-import type { Features } from 'trezor-connect';
 
 import LocalizableError from '../i18n/LocalizableError';
 import WalletTransaction from '../domain/WalletTransaction';
 import WalletAddress from '../domain/WalletAddress';
 import Wallet from '../domain/Wallet';
+import type { WalletHardwareInfo } from '../types/WalletType';
```
Comment (nicarq) [inline app/api/common.js]: `WalletHardwareInfo` is a little misleading because it holds information that Ledger doesn’t have

File: app/api/ada/hardwareWallet/newTransaction.js [hunk 3]
```diff
@@ -122,9 +127,9 @@ export async function newTrezorTransaction(
   }
 }
 
-function _derivePath(change: number, index: number): string {
+export function _derivePath(change: number, index: number): string {
   // Assumes this is only for Cardano and Web Yoroi (only one account).
-  return `${Config.trezor.DEFAULT_CARDANO_PATH}/${change}/${index}`;
+  return `${Config.wallets.CARDANO_FIRST_BIP_PATH}/${change}/${index}`;
```
Comment (nicarq) [inline app/api/ada/hardwareWallet/newTransaction.js]: @sarveshwar-singh FIRST_BIP_PATH doesn’t sound great. We should user the standardized naming https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki

File: app/api/ada/hardwareWallet/newTransaction.js [hunk 4]
```diff
@@ -154,3 +159,90 @@ function _outputAddressOrPath(out: TxOutput) {
...
+): Array<LedgerOutputTypeAddress | LedgerOutputTypeChange> {
+  return outputs.map(x => ({
+    amountStr: x.value.toString(),
+    ..._ledgerOutputAddressOrPath(x)
+  }));
+}
+
+function _ledgerOutputAddressOrPath(out: TxOutput) {
+  if (out.isChange) {
+    // It's a change output
```
Comment (nicarq) [inline app/api/ada/hardwareWallet/newTransaction.js]: I think we don’t need this comment because there is a line above that says isChange

File: app/domain/SignTx.js [hunk 5]
```diff
@@ -21,23 +24,23 @@ export type TrezorSignTxPayload = {
 
 export type BIP32Path = Array<number>;
 
-export type InputTypeUTxO = {|
+export type LedgerInputTypeUTxO = {
```
Comment (nicarq) [inline app/domain/SignTx.js]: @sarveshwar-singh I don’t understand why the file was changed to a generic name and this types were changed from generic to Ledger related 🤔

File: app/config.js [hunk 6]
```diff
@@ -6,17 +6,15 @@ export default {
     WALLET_RESTORED_NOTIFICATION_DURATION: 8,
     MAX_ALLOWED_UNUSED_ADDRESSES: 20,
     TRANSACTION_REQUEST_SIZE: 20,
-    WALLET_RECOVERY_PHRASE_WORD_COUNT: 12
+    WALLET_RECOVERY_PHRASE_WORD_COUNT: 12,
+    // https://github.com/trezor/connect/blob/develop/docs/methods/cardanoGetPublicKey.md
+    // https://cardanolaunch.com/assets/Ed25519_BIP.pdf    
+    CARDANO_FIRST_BIP_PATH: 'm/44\'/1815\'/0\'',
```
Comment (nicarq) [inline app/config.js]: @sarveshwar-singh we need to improve the name because it was really misleading

File: app/api/ada/index.js [hunk 7]
```diff
@@ -406,33 +406,35 @@ export default class AdaApi {
...
 
       if (error instanceof LocalizableError) {
         // we found it as a LocalizableError, so could throw it as it is.
         throw error;
-      } else {
-        // We don't know what the problem was so throw a generic error
-        throw new GenericApiError();
       }
+
+      // We don't know what the problem was so throw a generic error
```
Comment (nicarq) [inline app/api/ada/index.js]: @sarveshwar-singh why did you move it outside the else?

File: app/api/ada/index.js [hunk 8]
```diff
@@ -51,15 +51,13 @@ import {
 import type {
   TrezorSignTxPayload,
   LedgerSignTxPayload,
-} from '../../domain/TrezorSignTx';
+} from '../../domain/SignTx';
```
Comment (nicarq) [inline app/api/ada/index.js]: this could be `HWSignTx`. SignTx is too generic

File: app/api/ada/hardwareWallet/newTransaction.js [hunk 9]
```diff
@@ -154,3 +159,90 @@ function _outputAddressOrPath(out: TxOutput) {
...
+
+    Logger.debug('ledgerNewTransactions::_ledgerOutputAddressOrPath: ' +
+      `[WEIRD] Trezor got a change output without a full 'Ada Address': ${stringifyData(out)}`);
+  }
+
+  return { address58: out.address };
+}
+
+/** Send a transaction and save the new change address */
+export async function newLedgerTransaction(
```
Comment (nicarq) [inline app/api/ada/hardwareWallet/newTransaction.js]: @sarveshwar-singh this should be almost the same as Trezor, we shouldn’t create a new func

File: app/api/ada/hardwareWallet/newTransaction.js [hunk 10]
```diff
@@ -154,3 +159,90 @@ function _outputAddressOrPath(out: TxOutput) {
...
+function _transformToLedgerOutputs(
+  outputs: Array<TxOutput>
+): Array<LedgerOutputTypeAddress | LedgerOutputTypeChange> {
+  return outputs.map(x => ({
+    amountStr: x.value.toString(),
+    ..._ledgerOutputAddressOrPath(x)
+  }));
+}
+
+function _ledgerOutputAddressOrPath(out: TxOutput) {
```
Comment (nicarq) [inline app/api/ada/hardwareWallet/newTransaction.js]: @sarveshwar-singh missing return type and out is not a great name. This would be better with a type that is basically a string but it’s called Base58Address or something like that

File: app/api/ada/hardwareWallet/newTransaction.js (hunk 4 above)
Comment (sarveshwar-singh) [reply app/api/ada/hardwareWallet/newTransaction.js]: @nicarq sometimes i add comment to give some space to make it bit visibly clear and faster to understand.

File: app/api/ada/hardwareWallet/newTransaction.js (hunk 9 above)
Comment (sarveshwar-singh) [reply app/api/ada/hardwareWallet/newTransaction.js]: have plan to delete this function after checking.

File: app/api/ada/hardwareWallet/newTransaction.js (hunk 10 above)
Comment (sarveshwar-singh) [reply app/api/ada/hardwareWallet/newTransaction.js]: @nicarq 
As per my understanding there are 3 request:
1. missing return type
2. change the name to function parameters
3. [ Base58Address, is this about return type ?

For 1, 2 should i also change original functions which was written for Trezor Sign Tx part?

File: app/api/ada/index.js (hunk 7 above)
Comment (sarveshwar-singh) [reply app/api/ada/index.js]: @nicarq 
Yes, 
at some places it's inside and some places it's outside, outside looks cleaner.

File: app/domain/SignTx.js (hunk 5 above)
Comment (sarveshwar-singh) [reply app/domain/SignTx.js]: @nicarq 
Changing to generic because it store properties for Trezor and Ledger payload types.

 InputTypeUTxO -> LedgerInputTypeUTxO (to make it understandable that it belongs to ledger)

NOTE: 
1. For Ledger: this is temporary  (see the comment at top of the file), when ADA related `nmp` is out then these will no longer need.

2. For Trezor: these could not be exported from Trezor provided library ?

File: app/api/ada/hardwareWallet/newTransaction.js (hunk 10 above)
Comment (nicarq) [reply app/api/ada/hardwareWallet/newTransaction.js]: If you add it, now you own it lol 😄

File: app/api/ada/hardwareWallet/newTransaction.js (hunk 9 above)
Comment (nicarq) [reply app/api/ada/hardwareWallet/newTransaction.js]: please add TODOs

File: app/api/ada/index.js (hunk 7 above)
Comment (nicarq) [reply app/api/ada/index.js]: the convention that we are following is to use if { } else { }

File: app/domain/SignTx.js (hunk 5 above)
Comment (nicarq) [reply app/domain/SignTx.js]: `InputTypeUTxO -> LedgerInputTypeUTxO (to make it understandable that it belongs to ledger)`

This was already used by Trezor

Comment (sarveshwar-singh) [reply app/domain/SignTx.js]: @nicarq 
Sorry i did not understand your point here.

File: app/api/ada/index.js (hunk 7 above)
Comment (sarveshwar-singh) [reply app/api/ada/index.js]: @nicarq 
will revert and added issue for this PR's related non-conventional lines: https://github.com/Emurgo/yoroi-frontend/issues/308

File: app/api/ada/hardwareWallet/newTransaction.js (hunk 9 above)
Comment (sarveshwar-singh) [reply app/api/ada/hardwareWallet/newTransaction.js]: @nicarq 
I had already add it before but don't know why it's not visible in Github.
Although will add more details to the TODOs.

Comment (nicarq) [reply app/api/ada/hardwareWallet/newTransaction.js]: aha gotcha

File: app/utils/iframeHandler.js [hunk 11]
```diff
@@ -1,4 +1,25 @@
-// @flow 
+// @flow
+import { LedgerBridge } from 'yoroi-extension-ledger-bridge';
+
+export async function prepareLedgerBridger(ledgerBridge: LedgerBridge): Promise<void> {
```
Comment (nicarq) [inline app/utils/iframeHandler.js]: this is good. I would like to handle a timeout case as welll. What do you think of something like this:

```
_wait = async ms => new Promise((resolve) => setTimeout(resolve, ms));

export async function waitIFrameLoaded(name: string): Promise<void> {
  if (ledgerBridge == null) {
    throw new Error(`Error: LedgerBridge is undefined`);
  }

  const promise = new Promise((resolve, reject) => {
    const start = new Date();
    const maxWait = 10000; // 10 seconds

    while (iframeLoaded) {
      _wait(500);
      iframeLoaded = ledgerBridge.isReady;

      if ((new Date().getTime() - start) > maxWait) {
        reject(new Error(`Time out. IFrame ${name} wasn’t able to load in 10 seconds.`));
      }
    }

    resolve();
  });

  return promise;
}
```

File: .eslintignore [hunk 12]
```diff
@@ -1,3 +1,4 @@
 /js-cardano-wasm
 /dll
-/chrome/3rd-party/trezor
\ No newline at end of file
+/chrome/3rd-party/trezor
+app/api/ada/hardwareWallet/utils.js
```
Comment (sarveshwar-singh) [inline .eslintignore]: since, this file is replaced by ledger bridge package so now this line can be deleted.

File: app/stores/ada/TrezorConnectStore.js [hunk 13]
```diff
@@ -375,6 +381,29 @@ export default class TrezorConnectStore extends Store {
...
+        publicMasterKey,
+        vendor: deviceFeatures.vendor,
+        model: deviceFeatures.model,
+        deviceId: deviceFeatures.device_id,
+        label: deviceFeatures.label,
+        majorVersion: deviceFeatures.major_version,
+        minorVersion: deviceFeatures.minor_version,
+        patchVersion: deviceFeatures.patch_version,
+        language: deviceFeatures.language,
+        chainCodeHex: '',
```
Comment (nicarq) [inline app/stores/ada/TrezorConnectStore.js]: @sarveshwar-singh what is this for?

File: app/containers/wallet/dialogs/WalletLedgerConnectDialogContainer.js [hunk 14]
```diff
@@ -0,0 +1,84 @@
+// @flow
+import React, { Component } from 'react';
+import { observer } from 'mobx-react';
+
+import environment from '../../../environment';
+import type { InjectedDialogContainerProps } from '../../../types/injectedPropsType';
```
Comment (nicarq) [inline app/containers/wallet/dialogs/WalletLedgerConnectDialogContainer.js]:
```suggestion
import type { InjectedDialogContainerProps as Props } from '../../../types/injectedPropsType';
```

Comment (nicarq) [reply app/containers/wallet/dialogs/WalletLedgerConnectDialogContainer.js]: and delete `type Props = InjectedDialogContainerProps;`

File: app/components/wallet/hwConnect/ledger/ConnectDialog.js [hunk 15]
```diff
@@ -30,17 +30,12 @@ const connectStartGIF = connectLoadGIF;
 const messages = defineMessages({
   connectIntroTextLine1: {
     id: 'wallet.connect.ledger.dialog.step.connect.introText.line.1',
-    defaultMessage: '!!!After connecting your Trezor device to the computer press the Connect button.',
+    defaultMessage: '!!!After connecting your Ledger device to your computer\'s USB port, press the Connect button.',
```
Comment (nicarq) [inline app/components/wallet/hwConnect/ledger/ConnectDialog.js]: we should use `’` instead of `'`

File: app/containers/wallet/WalletSendPage.js [hunk 16]
```diff
@@ -25,36 +25,36 @@ import HWSendConfirmationDialog from '../../components/wallet/send/HWSendConfirm
 const messagesLedger: MessageDescriptorMap = defineMessages({
   infoLine1: {
     id: 'wallet.send.ledger.confirmationDialog.info.line.1',
-    defaultMessage: '!!!After connecting your Trezor device to your computer, press the Send using Trezor button.',
-    description: 'Informative message line 1 in the wallet trezor send confirmation dialog.'
+    defaultMessage: '!!!After connecting your Ledger device to your computer\'s USB port, press the Send using Ledger button.',
```
Comment (nicarq) [inline app/containers/wallet/WalletSendPage.js]: we should use `’` instead of `'`

File: app/components/wallet/hwConnect/ledger/AboutDialog.js [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+            <li key="1">
+              {intl.formatMessage(messages.aboutPrerequisite1Part1)}
+              <a target="_blank" rel="noopener noreferrer" href={intl.formatMessage(messages.aboutPrerequisite1Part2Link)}>
+                {intl.formatMessage(messages.aboutPrerequisite1Part2LinkText)}
+                <SvgInline svg={externalLinkSVG} cleanup={['title']} />
+              </a>
+              {intl.formatMessage(messages.aboutPrerequisite1Part3)}
+            </li>
+            <li key="2">
+              <a target="_blank" rel="noopener noreferrer" href={intl.formatMessage(messages.aboutPrerequisite2Part1Link)}>
```
[the remark this answers is not in the document]
Comment (nicarq) [reply app/components/wallet/hwConnect/ledger/AboutDialog.js]: good catch, Seba!. we shouldn’t

[the remark this answers is not in the document]
Comment (nicarq) [reply app/components/wallet/hwConnect/ledger/AboutDialog.js]: maybe instead we could just add a redirect using our DNS, so there is a possibility to know how many people are clicking them without giving up on the referrer stuff

File: app/api/ada/hardwareWallet/newTransaction.js [hunk 18]
```diff
@@ -0,0 +1,468 @@
...
+  const trezorOutputs = _generateTrezorOutputs(txExt.outputs);
+
+  // Transactions
+  const txsBodies = await txsBodiesForInputs(txExt.inputs);
+
+  // Network
+  const network = CONFIG.network.trezorNetwork;
+
+  return {
+    // TODO [TREZOR] change (network -> protocol_magic)
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/api/ada/hardwareWallet/newTransaction.js]: yes, deleted

File: app/api/ada/hardwareWallet/newTransaction.js [hunk 19]
```diff
@@ -0,0 +1,468 @@
...
+      return { path: _derivePathAsBIP32Path(fullAddress.change, fullAddress.index) };
+    }
+    Logger.debug(`newTransaction::_ledgerOutputAddressOrPath:[WEIRD] Ledger got a change output without a full 'Ada Address': ${stringifyData(txOutput)}`);
+  }
+
+  return { address58: txOutput.address };
+}
+
+/** Send a transaction and save the new change address */
+// TODO: Change name. This does way moer than ”broadcast”
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/api/ada/hardwareWallet/newTransaction.js]: DONE

File: app/stores/ada/LedgerConnectStore.js [hunk 20]
```diff
@@ -0,0 +1,389 @@
...
+    return Config.wallets.hardwareWallet.ledgerNanoS.DEFAULT_WALLET_NAME;
+  }
+
+  get isActionProcessing(): boolean {
+    return this.progressInfo.stepState === StepState.PROCESS;
+  }
+  // =================== VIEW RELATED =================== //
+
+  // =================== API RELATED =================== //
+  createHWRequest: LocalizedRequest<any> =
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: seems like Github source was outdated.

File: app/stores/ada/LedgerConnectStore.js [hunk 21]
```diff
@@ -0,0 +1,389 @@
...
+        // https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#examples
+        Logger.debug(stringifyData(CARDANO_FIRST_ACCOUNT_SUB_PATH));
+
+        // get Cardano's first account's
+        // i.e hdPath = [2147483692, 2147485463, 2147483648]
+        const extendedPublicKeyResp: GetExtendedPublicKeyResponse
+          = await ledgerBridge.getExtendedPublicKey(CARDANO_FIRST_ACCOUNT_SUB_PATH);
+        Logger.debug(stringifyData(extendedPublicKeyResp));
+
+        // 🔻TOD0: Delete. Debug purposes
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: TODO

File: app/stores/ada/LedgerConnectStore.js [hunk 22]
```diff
@@ -0,0 +1,389 @@
...
+  ): boolean => {
+    if (versionResp == null) {
+      throw new UnexpectedError();
+    }
+
+    if (extendedPublicKeyResp == null) {
+      throw new UnexpectedError();
+    }
+
+    // TODO [LEDGER]: check validity
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: seems like Github source was outdated. DONE

File: app/stores/ada/LedgerConnectStore.js [hunk 23]
```diff
@@ -0,0 +1,389 @@
...
+  _convertToLocalizableError = (error: any): LocalizableError => {
+    let localizableError: ?LocalizableError = null;
+
+    console.log(JSON.stringify(error, null, 2));
+
+    if (error instanceof LocalizableError) {
+      // It means some API Error has been thrown
+      localizableError = error;
+    } else if (error && error.message) {
+      // // Trezor device related error happend, convert then to LocalizableError
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: that block was TODO, fixed now.

File: app/stores/ada/LedgerConnectStore.js [hunk 24]
```diff
@@ -0,0 +1,389 @@
...
+    Logger.debug('LedgerConnectStore::_onSaveSucess loading wallet data');
+    wallets.refreshWalletsData();
+
+    // Load the Yoroi with Trezor Icon
+    this.stores.topbar.initCategories();
+
+    // show success notification
+    wallets.showLedgerNanoSWalletIntegratedNotification();
+
+    // TODO: [LEDGER] not sure if it actully destroying this Store ??
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: removed comment,
`teardown` removes `reactions` (if registered)

File: app/stores/ada/LedgerConnectStore.js [hunk 25]
```diff
@@ -0,0 +1,389 @@
...
+
+    // goto the wallet transactions page
+    Logger.debug('LedgerConnectStore::_onSaveSucess setting new walles as active wallet');
+    wallets.goToWalletRoute(ledgerWallet.id);
+
+    // fetch its data
+    Logger.debug('LedgerConnectStore::_onSaveSucess loading wallet data');
+    wallets.refreshWalletsData();
+
+    // Load the Yoroi with Trezor Icon
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: DONE

File: app/stores/ada/LedgerSendStore.js [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+  ): Promise<void> => {
+
+    const reqParams: BroadcastLedgerSignedTxRequest = {
+      ledgerSignTxResp,
+      changeAdaAddr: createLedgerSignTxDataResp.changeAddress,
+      unsignedTx,
+      txExt: createLedgerSignTxDataResp.txExt
+    };
+
+    // TODO: [TREZOR] add error check
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerSendStore.js]: removed comment,
is case of error API will throw Error

File: app/stores/ada/LedgerSendStore.js [hunk 27]
```diff
@@ -0,0 +1,239 @@
...
+  /** Converts error(from API or Trezor API) to LocalizableError */
+  _convertToLocalizableError = (error: any): LocalizableError => {
+    let localizableError: ?LocalizableError = null;
+
+    if (error instanceof LocalizableError) {
+      // It means some API Error has been thrown
+      localizableError = error;
+    } else if (error && error.message) {
+      // Trezor device related error happend, convert then to LocalizableError
+      // TODO: [TREZOR] check for device not supported if needed
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerSendStore.js]: yes, no need of TODO, fixed

File: app/stores/ada/LedgerConnectStore.js [hunk 28]
```diff
@@ -0,0 +1,389 @@
...
+
+      const reqParams = this._prepareCreateHWReqParams(walletName);
+      const ledgerWallet: CreateHardwareWalletResponse =
+        await this.createHWRequest.execute(reqParams).promise;
+
+      if (ledgerWallet) {
+        await this._onSaveSucess(ledgerWallet);
+      } else {
+        // this Error will be converted to LocalizableError()
+        throw new Error();
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: i think `if` check is not needed.

[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: I think `if` block is not needed.

File: app/stores/ada/LedgerConnectStore.js [hunk 29]
```diff
@@ -0,0 +1,389 @@
...
+
+      if (error instanceof CheckAdressesInUseApiError) {
+        // redirecting CheckAdressesInUseApiError -> hwConnectDialogSaveError101
+        // because for user hwConnectDialogSaveError101 is more meaningful in this context
+        this.error = new LocalizableError(globalMessages.hwConnectDialogSaveError101);
+      } else if (error instanceof LocalizableError) {
+        this.error = error;
+      } else {
+        // some unknow error
+        this.error = new UnexpectedError();
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: that is not swallow, it has fixed message inside it.
`UnexpectedError` is a `LocalizableError` for UI display (in case we are not sure which error happened)

debugging is done in upper block.
`Logger.error(`LedgerConnectStore::_saveHW::error ${stringifyError(error)}`);`

File: app/stores/ada/LedgerConnectStore.js [hunk 30]
```diff
@@ -0,0 +1,389 @@
...
+      this.createHWRequest.reset();
+      this._setIsCreateHWActive(false);
+    }
+  };
+
+  _prepareCreateHWReqParams = (walletName: string): CreateHardwareWalletRequest => {
+    if (this.hwDeviceInfo == null
+      || this.hwDeviceInfo.publicMasterKey == null
+      || this.hwDeviceInfo.hwFeatures == null) {
+      throw new UnexpectedError();
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerConnectStore.js]: as `UnexpectedError` has fixed message, for debugging we can convert that to  ` Error('Ledger device hardware info not valid')`

File: app/stores/ada/TrezorConnectStore.js [hunk 31]
```diff
@@ -321,58 +324,62 @@ export default class TrezorConnectStore extends Store {
...
-        this.error = new LocalizableError(messages.saveError101);
+        // redirecting CheckAdressesInUseApiError -> hwConnectDialogSaveError101
+        // because for user hwConnectDialogSaveError101 is more meaningful in this context
+        this.error = new LocalizableError(globalMessages.hwConnectDialogSaveError101);
       } else if (error instanceof LocalizableError) {
         this.error = error;
       } else {
         // some unknow error
-        this.error = new LocalizableError(messages.error999);
+        this.error = new UnexpectedError();
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/TrezorConnectStore.js]: that is not swallow, it has fixed message inside it.
UnexpectedError is a LocalizableError for UI display (in case we are not sure which error happened)

File: app/stores/ada/TrezorConnectStore.js [hunk 32]
```diff
@@ -169,54 +156,125 @@ export default class TrezorConnectStore extends Store {
...
+
+  _handleConnectError = (error): void => {
+    Logger.error(`TrezorConnectStore::_handleConnectError ${stringifyError(error)}`);
+
+    this.hwDeviceInfo = undefined;
+
+    if (error instanceof LocalizableError) {
+      this.error = error;
+    } else {
+      this.error = new UnexpectedError();
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/TrezorConnectStore.js]: that is not swallow, it has fixed message inside it.
UnexpectedError is a LocalizableError for UI display (in case we are not sure which error happened)

File: app/stores/ada/LedgerSendStore.js [hunk 33]
```diff
@@ -0,0 +1,239 @@
...
+      if (this.ledgerBridge) {
+        // Since this.ledgerBridge is undefinable flow need to know that it's a LedgerBridge
+        const ledgerBridge: LedgerBridge = this.ledgerBridge;
+
+        const ledgerSignTxDataResp: CreateLedgerSignTxDataResponse =
+          await this.createLedgerSignTxDataRequest.execute(params).promise;
+
+        await prepareLedgerBridger(ledgerBridge);
+
+        // TODO: Fix types
```
[the remark this answers is not in the document]
Comment (sarveshwar-singh) [reply app/stores/ada/LedgerSendStore.js]: seems like Github source was outdated. DONE
